fuse_variables/src/orientation_2d_stamped.cpp

```diff
@@ -31,14 +31,82 @@
  *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
  *  POSSIBILITY OF SUCH DAMAGE.
  */
+#include <fuse_variables/orientation_2d_stamped.h>
+
 #include <fuse_core/local_parameterization.h>
 #include <fuse_core/uuid.h>
-#include <fuse_variables/orientation_2d_stamped.h>
+#include <ros/time.h>
+
+#include <cmath>
 
 
 namespace fuse_variables
 {
 
+/**
+ * @brief A LocalParameterization class for 2D Orientations.
+ *
+ * 2D orientations add and subtract in the "usual" way, except for the 2*pi rollover issue. This local parameterization
+ * handles the rollover. Because the Jacobians for this parameterization are always identity, we implement this
+ * parameterization with "analytic" derivatives, instead of using the Ceres's autodiff system.
+ */
+class Orientation2DLocalParameterization : public fuse_core::LocalParameterization
+{
+public:
+  // Define some necessary variations of PI
+  static constexpr double M_TWO_PI = 2 * M_PI;
+
+  int GlobalSize() const override
+  {
+    return 1;
+  }
+
+  int LocalSize() const override
+  {
+    return 1;
+  }
+
+  bool Plus(
+    const double* x,
+    const double* delta,
+    double* x_plus_delta) const override
+  {
+    // Compute the angle increment as a linear update
+    x_plus_delta[0] = x[0] + delta[0];
+    // Then handle the 2*Pi roll-over
+    x_plus_delta[0] -= M_TWO_PI * std::floor((x_plus_delta[0] + M_PI) / M_TWO_PI);
+    return true;
+  }
+
+  bool ComputeJacobian(
+    const double* x,
+    double* jacobian) const override
+  {
+    jacobian[0] = 1.0;
+    return true;
+  }
+
+  bool Minus(
+    const double* x1,
+    const double* x2,
+    double* delta) const override
+  {
+    // Compute the difference from x2 to x1
+    delta[0] = x2[0] - x1[0];
+    // Then handle the 2*Pi roll-over
+    delta[0] -= M_TWO_PI * std::floor((delta[0] + M_PI) / M_TWO_PI);
```

Comment:
Could we make this logic a templated method somewhere?

```
delta[0] = normalizeRotation(x2[0] - x1[0]);
```

Reply:
We seem to have a slightly annoying problem; we have several utility functions defined in two different places:
https://github.com/locusrobotics/fuse/blob/devel/fuse_constraints/include/fuse_constraints/util.h
https://github.com/locusrobotics/fuse/blob/devel/fuse_variables/include/fuse_variables/util.h

For the moment, I'm making it worse by copying the missing functions from fuse_constraints/util.h to fuse_variables/util.h.

Maybe more them to fuse_core? Even though fuse_core doesn't use them?

Reply:
Moved util.h to fuse_core